### ISSUE TITLE
Recursiveness of `vec_cast()` with list `x`

### ISSUE THREAD

**Issue by DavisVaughan** [r-lib/vctrs#738]
Because of the way `vec_list_cast()` works, this recursive casting idea is allowed. @lionel- pointed out that the strict semantics won't have this problem because even `vec_cast(list(1), integer())` won't be allowed, but if there is some version of `vec_force()` that sticks around that can do list -> integer conversion then it should handle this properly.

```r
library(vctrs)

vec_cast(list(1), integer())
#> [1] 1

# should error here
vec_cast(list(list(1)), integer())
#> [1] 1
vec_cast(list(list(list(1))), integer())
#> [1] 1
vec_cast(list(list(list(list(1)))), integer())
#> [1] 1
```

**Comment by DavisVaughan** [comment]
Maybe we should be using `vec_coercible_cast()` here?

https://github.com/r-lib/vctrs/blob/ab84679b816b903cbafa9aa042ec1c9c98d385a5/R/cast-list.R#L33

**Comment by DavisVaughan** [comment]
Fixed by #963 and all of these error now
